tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11400,16 +11400,78 @@ bool hlsl::DiagnoseNodeStructArgument(Sema *self, TemplateArgumentLoc ArgLoc,
   }
 }
 
+bool IsExported(Sema *self, clang::FunctionDecl *FD) {
+  if (FD->hasAttr<HLSLShaderAttr>()) {
+    return true;
+  }
+
+  bool isMarkedStatic = false;
+  bool isMarkedExport = false;
+
+  if (auto cDecl = dyn_cast<CXXMethodDecl>(FD)) {
+    if (cDecl->isStatic()) {
+      isMarkedStatic = true;
+    }
+  }
+
+  if (FD->hasAttr<HLSLExportAttr>()) {
+    isMarkedExport = true;
+  }
+
+  if (isMarkedStatic && isMarkedExport) {
+    self->Diag(FD->getLocation(), diag::err_hlsl_varmodifiersna) << "static"
+                                                                 << "export"
+                                                                 << "function";
+    return false;
+  }
+
+  Linkage L = FD->getLinkageAndVisibility().getLinkage();
+
+  // case 1 requires special assignments to linkage
+  if (L != ExternalLinkage && L != InternalLinkage) {
+    // the condition below implies the shader stage is lib_6_x
+    if (self->getLangOpts().IsHLSLLibrary) {
+      L = ExternalLinkage;
+    } else {
+      L = InternalLinkage;
+    }
+  }
+
+  // now case 2 and 3 can apply and we can determine
+  // whether the function is exported.
+  if (L == InternalLinkage && isMarkedExport) {
+    return true;
+  }
+
+  if (L == ExternalLinkage && isMarkedStatic) {
+    return false;
+  }
+
+  if (L == InternalLinkage) {
+    return false;
+  }
+
+  if (L == ExternalLinkage) {
+    return true;
+  }
+  // unreachable but necessary to prevent warnings
+  return true;
+}
+
 // validates that for every function in the translation unit, if it
 // references a patch constant function, there exists a function declaration
 // that could serve as a candidate to that patch constant function.
 void ValidatePatchConstantFunctionsExist(clang::Sema *self) {
+
   for (auto decl : self->getASTContext().getTranslationUnitDecl()->decls()) {
     // TODO: improve condition so that only exported functions are checked,
     // instead of all functions. Issue: #5857
     if (FunctionDecl *FD = dyn_cast<FunctionDecl>(decl)) {
       // If there is no patch constant function, then we don't need to validate
       // anything.
+      if (!IsExported(self, FD)) {
```

Comment:
What if a function is not exported, but used by exported function?

Reply:
My understanding during implementation was that exported functions and all functions reachable from exported functions should be checked for recursion. I know for sure that exported functions should be checked for recursion. So I do think everything reachable should be checked too because otherwise we would be missing a diagnostic on actual recursion that takes place during execution, right?